### ISSUE TITLE
BUG: TreeSHAP Interventional explanations segmentation fault 

### ISSUE THREAD

**Issue by amir-rahnama** [shap/shap#3486]
### Issue Description

When you explain RandomForest models (and sometimes even GradientBoostingTrees with Sklearn) using TreeSHAP explainer, tree-shap script breaks. 

I was assuming that this was fixed in [BUG: Interventional TreeSHAP failing for large depth tree-based models](https://github.com/shap/shap/issues/3267), but this is still recurring in SHAP version 0.44.1 and even 0.43.0. 

Colab link: https://colab.research.google.com/drive/1bAH-4WclIJQPvPwoY3cZ-UZu8MAccaTO?usp=sharing

### Minimal Reproducible Example

```python
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_moons, make_circles, make_classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import KBinsDiscretizer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.datasets import fetch_openml
from joblib import dump, load
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
import seaborn as sns
import argparse
import numpy as np
import pandas as pd
import sys
import os
import pickle
import scipy
from sklearn.preprocessing import StandardScaler
import sklearn
import collections
import shap

random_state = 10
np.random.seed(random_state)

dset_id = 293

ames = fetch_openml(data_id = dset_id, as_frame='auto')
random_state = 10
ames.target[ames.target == '-1'] = 0
ames.target[ames.target == '1'] = 1

target = ames.target.astype(int)
ames.data = ames.data.toarray()

if np.sum(np.isnan(ames.data)):
    feat_col_means = np.nanmean(ames.data, axis=0)
    ames.data = np.where(np.isnan(ames.data), feat_col_means, ames.data)

X_train, X_test, y_train, y_test = train_test_split(ames.data, target, test_size=0.33, random_state=10)

model_name = 'rf'

if model_name == 'rf':
    model = RandomForestClassifier(random_state=random_state)
else: 
    model = GradientBoostingClassifier(random_state=random_state)

model.fit(X_train, y_train)

shap_explainer = shap.TreeExplainer(model, X_train, 
                                    feature_perturbation="interventional",
                                    model_output='probability')

def tree_shap_exp(instances, x_train, model_obj, x_test):
    shap_values = shap_explainer.shap_values(instances, check_additivity=False)
    shap_values = np.array(shap_values)

    return shap_values

res = tree_shap_exp(X_test[:10], X_train, model, X_test)
```

### Traceback

```shell
Segmentation fault (core dumped)
```

### Expected Behavior

I expect TreeSHAP to work since neither the dataset nor the tree model is large enough not to fit into the memory. 

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

* SHAP == 0.44.1 or SHAP == 0.43.0
* Sklearn == 1.4.0 or 1.3.0 
* OS  Ubuntu 20.04.6 LTS or MAC M1 Montrey 
* Available Memory on Server: 
              total        used        free      shared  buff/cache   available
Mem:      196671072    20887784    62542380      123692   113240908   174204456
Swap:       8388604     4712112     3676492

**Comment by amir-rahnama** [comment]
Here is the crash logs on Google Colab:
<img width="493" alt="Screenshot 2024-02-06 at 11 59 27" src="https://github.com/shap/shap/assets/6009583/6231a5b8-8cb6-4c22-9c49-f89a93086804">

**Comment by amir-rahnama** [comment]
**UPDATE**: If you set max_depth=30 for training the RF, the issue will go be fixed. 

Possible other issues related to this: 

* https://github.com/shap/shap/issues/3371
* https://github.com/shap/shap/issues/2076
* https://github.com/shap/shap/issues/3390

**Comment by CloseChoice** [comment]
Thanks for your report. I remember that we had the problem previously. Seems like a deep bug in our C code

**Comment by stergioa** [comment]
> Thanks for your report. I remember that we had the problem previously. Seems like a deep bug in our C code

Any fix?

**Comment by CloseChoice** [comment]
@stergioa thanks for the gentle push. AFAIK nobody is working on this. Just checked the example and it doesn't throw an error on my Windows PC. We would certainly need one first before we can start working on this.

**Comment by stoensin** [comment]
model: CatBoostClassifier with multiclass(3 class) loss function, with string type cat_features;
total feature num: 50;
while model iterations>3000+, when call shap.Explainer(class_model) will definitely crash;

**Comment by amir-rahnama** [comment]
> @stergioa thanks for the gentle push. AFAIK nobody is working on this. Just checked the example and it doesn't throw an error on my Windows PC. We would certainly need one first before we can start working on this.

It's highly dependent on your hardware and configuration. Try to increase the model size to challenge your hardware and I assume it will throw the error :)

**Comment by tudstudent** [comment]
I tested this issue (interventional TreeSHAP segfault with large RandomForest models) against the current master branch, which includes several interventional TreeSHAP fixes from PR #4263:

**Test setup:**
- 500 samples, 50 features, sparse data, no max_depth limit
- RandomForestClassifier with 100 estimators (max tree depth: 19)
- Interventional mode with 100 background samples

**Results:**
- SHAP values: computed successfully, no NaN, no Inf, no segfault
- Interaction values: computed successfully, no NaN, no Inf, no segfault

The segfault may have been related to memory layout issues in the C++ code that have since been addressed. In PR #4263, we refactored `from_flag` out of the shared `Node` struct into per-thread arrays, which eliminates a potential data race that could have caused memory corruption.

If you can still reproduce this on the latest master, please share the exact dataset/model configuration.

**Comment by clark874** [comment]
Root cause identified: the independent TreeSHAP C++ path (`tree_shap_indep` and the `Node` struct in `shap/cext/tree_shap.h`) stores tree node indices in 16-bit `short`s. Any single tree with more than 2**15 - 1 = 32767 nodes overflows these fields, after which the traversal reads/writes `mytree[negative]`, corrupting the heap and segfaulting inside `shap_values()`. This is independent of forest size, tree depth, input dtype, and sklearn/shap version (reproduced on 0.43–0.52), and it affects every model format when `feature_perturbation="interventional"` — deep random forests (30–40k nodes per tree) trigger it reliably. `tree_path_dependent` uses 32-bit indices and is unaffected.

Minimal repro: build one tree with ≥32768 nodes (dict model format works, ~1s), explain with `feature_perturbation="interventional"` → deterministic SIGSEGV.

Fix and regression test: #5163 — widens the affected fields to `int` (4 lines) plus a test that constructs a >2**15-node tree and asserts SHAP additivity. Verified: crashes before the patch, passes after; a real-world 500-tree / 19.2M-node RandomForest that segfaulted on 0.52 now computes interventional SHAP correctly.
